src/krb5_plugin/passkey/passkey_clpreauth.c

```diff
@@ -52,14 +52,58 @@ sss_passkeycl_prep_questions(krb5_context context,
 {
     struct sss_passkey_message *message;
     char *question = NULL;
+    char **realms = NULL;
     krb5_error_code ret;
+    size_t r = 0;
 
     message = sss_passkey_message_decode_padata(pa_data);
     if (message == NULL) {
         ret = ENOMEM;
         goto done;
     }
 
+    if (message->data.challenge->domain == NULL) {
+        ret = KRB5KDC_ERR_PREAUTH_FAILED;
+        goto done;
+    }
+
+    /* Find a realm that matches the domain from the challenge */
+    ret = krb5_get_host_realm(context,
+                              message->data.challenge->domain,
+                              &realms);
+    if (ret || ((realms == NULL) || (realms[0] == NULL))) {
+        ret = KRB5KDC_ERR_PREAUTH_FAILED;
+        goto done;
+    }
+
+    /* Do explicit check for the challenge domain in case
+     * we've got back a referral (empty) realm */
+    if (strlen(realms[0]) == strlen(KRB5_REFERRAL_REALM)) {
+        ret = strncasecmp(message->data.challenge->domain,
+                          request->server->realm.data,
+                          request->server->realm.length);
+        if (ret == 0) {
```

Comment:
Hi,

`ret == 0` means success in this case and not failure.

bye,
Sumit

Reply:
Thank you, changed.